utils/docs/gcheck-library.1.in
.\" groff -man -Tascii foo.1
.TH gcheck-library 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gcheck-library \- A script to assist in maintaining symbol libraries

.SH SYNOPSIS
.B gcheck-library
.RI [ OPTIONS ] [ DIRECTIVES [ ARGUMENTS ] ] [ LIBRARY ]
.SH DESCRIPTION
.B gcheck-library
Detect non complient symbols files in gEDA symbol libraries.
.SH QUICK
cd /symbols-dir, type gcheck-library <ENTER>.
.SH INFORMATION
Use of the following options displays information related to gcheck-library
but does not check anything:
.IP "-h, --help"
Show information about the advanced usage of this command.
.IP "-u, --usage"
Displays examples on usage"
.IP "--version"
Displays software version information"
.SH OPTIONS
.IP "-n, --no_report"
do not perform report statistics.'
.IP "-r, --report"
report statistic only, do not perform checking. If the no report options is used then the exit code will be the number of symbol files or -1 if an error occurred.'
.SH DIRECTIVES
.IP "-v, --verbose"
Include additional information during processing.
Directives affecting the mode and operation of the gcheck-library utility.
.IP "-a, --all"
file search pattern is changed from *.sym to *'.
.IP "-c, --clip"
Performs advanced checks on geda-clib.scm. This directive is not intended for regular users. When enabled, symbol files are
not checked, instead, gcheck-library will perform checks related to geda-clib.scm in two stages. The first check is for the
syntax of geda-clib.scm. The gaf utility's shell feature is utilized to load the file in guile, this is a pass or fail check.
The results are reported. The second stage checks for references to the names of symbol folders in geda-clib.scm, which
uses the "component-library" function to load the main library, so any folder names reported as being unreferenced will not
appear in the component selection dialog in gschem. Note that this directive can only be used in the installed location
of the geda-clib.scm or in the source directory containing the geda-clib.scm.
.IP "-d, --docs"
Check validity of URL referenced by documentation attributes.
.IP "-f, --filter"
The Filter directive is intended to over-ride the internal default symbol file type.
.SH DIAGNOSTICS
Note: arguments are case sensitive. See documentation for more detailed information. [dirname] &2>result.txt.
.SH HOMEPAGE
http://www.geda-project.org
.SH BUGS
You can report bugs at the Launchpad site of gcheck-library:
https://bugs.launchpad.net/geda/gcheck-library/+filebug
.SH AUTHOR
Wiley Edward Hill < wileyhill dot gmail dot com >
.SH SEE ALSO
.BR gcheck-library (1)

.SH SEE ALSO
.BR gcheck-library (1),
.BR gsymcheck (1)

.SH COPYRIGHT
.nf
Copyright \(co 2012-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the "COPYING" file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it. There is NO
WARRANTY; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, to
the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 04/01/19 Rampage Edition

* Expanded support for internationalization (i18n).

* All utilities now support internationalization (i18n).

* The visibility of toolbar tooltips can now be control using menu options.

* The context menu available when creating Polygon objects now includes
  an UNDO option allowing users to undo the previous section and resume
  path mode from the previous node.

* Scripting to support printing schematic documents from the command-line
  using gschem has been enhanced to support postscript or pdf output.
  Previously the default was changed from postscript to pdf. Postscript
  outputs could still be produced using menu options available in gschem.
  The output type is now automatically detected when using the command
  line interface by the file extension of the output file name. The
  default remains pdf if the output name is not given.

* Configuration function output-type has been renamed output-type.

* Libgedacairo now uses round end-caps when drawing buses, resulting in
  smoother transitions and improved appearance.

* A new keyword option hide-columns has been added to gattrib to allow
  user to designate a list attributes to be hidden in the GUI be default.
  A new Column Visibility dialog has been added to support user management
  of visible columns.

* A new path backend has been added to the gaf tool to allow scripting
  lanauages other than scheme to retrieve path information from libgeda.

* Improved appearence of various dialogs in gschem.

* gschem now logs exiting of the schematic editor if log-time is enabled.

* gschem now reports failure to attach attributes that aleady attached.

* gschem text properties dialog now supports operations to manipulate
  text case.

* gschem translate dialog now allows for translation of only selected
  objects. Additionally, the translate offset entry now accepts X and
  Y coordinates.

* Auxiliary buffers can now be explicitly cleared using a right context
  menu on the corresponding paste menu item.

* The suite now include template documents install to the user's template
  folder so the user can create new schematic and symbol documents from
  the system's context menu.

* Issues resolved:

* Libgedacairo has been enhanced to resolve issues with the end points of
  arcs shifting at different zoom levels. The shifting sometimes resulted
  in misalignment to adjoining lines at higher magnification levels.

* Extensive revisions were made to the Rampage Edition to build the suite
  on MinGW platforms. The Rampage Edition is now fully functional under
  native Win32 host systems including i18n support.

* The Rename Session dialog now fills in the previous session name,
  previously to dialog presented the user with an empty name field.

* The locked bit is not preserved when copying components in gschem.

* gschem now restarts text search when users change text in the Find Text
  dialog and click "find" or hits ENTER instead of closing the dialog.

* gschem now raises the main window when interacting with the Array Dialog.
  Previously, user could selected other application windows and return to
  the dialog and select the button that lowered the dialog but did not
  ensure the main window was in the foreground.

* An issue was resolved with controls not being initially enabled on the
  Attributes tab of the Preferences Dialog if the Attributes tab was the
  "saved" initial tab in gschem. Switching tabs or the view list controls
  re-enabled the disabled controls. The saved tab is now restored after
  settings are uploaded to the Preferences Dialog and this is reflected
  in the initial state of controls on the Attributes tab.

* An issue was resolved with resetting component-dialog-attributes when
  a second list was encountered in rc files. If a second attributes list
  was processed gschem would fault when the list was accessed or exiting
  gschem.

* The drc2 back end no longer reports duplicate slots for on components
  that do not have slots. This generally occurred when components have
  duplicate reference designators because these would typically show up
  using the same slot number.

* Libgeda no longer reports an oddity regarding symbol versions if the
  oddity was due to the symbol file being missing.

* The scheme file geda-deprecated-config.scm is no longer installed.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.


utils/scripts/gcheck-library.sh
#!/bin/bash

# Author: Wiley Edward Hill
# Date: 10/20/12
#
# gEDA - GPL Electronic Design Automation
# gcheck-library - gEDA Symbol Library Batch Processor
#
# Copyright (C) 2012-2015 Wiley Edward Hill <wileyhill@gmail.com>
#
#------------------------------------------------------------------
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#------------------------------------------------------------------

VER=1.2.0

VERBOSE=false
QUIET=false
REPORT=true
CHECK_CLIB=false
CHECK_DOCS=false
FILTER="*.sym"
PROGRAM1="gsymcheck"
PROGRAM2="gaf"
PROGRAM3="curl"
OPT=""
FINAL_ARGUMENT="${@: -1}"

# directory names to be excluded when checking geda-clib.scm
EXCLUDE='! -name doc ! -name src ! -name model'

ERR_FILE_NOT_FOUND=2
ERR_BAD_ARGS=65

do_show_help()
{
   case $1 in
         "1") # Quick help
              echo Usage:   `basename $0` '[-h] | [-v] | [-r] | [-a,d,e,n] [-f <pattern>] [ Library ]' ;;
         "2") # Basic help
              echo
              echo "A script to assist in maintaining symbol libraries"
              echo Usage:   `basename $0` '[-options] [-d] [-f <pattern>] [ Library ]'
              echo
              echo "Use the following options to control behavior:"
              echo
              echo   '  -a | --all       file pattern is changed from *.sym to *'
              echo   '  -c | --clip      Check geda-clib.scm for errors and references'
              echo   '  -d | --docs      Check documentation references'
              echo   '  -f | --filter <pattern>   filter pattern is changed to pattern'
              echo   '  -n | --no_report do not perform report statistic'
              echo   '  -r | --report    report statistic only, do not perform checking'
              echo   ''
              echo   '  -h | --help      show this information'
              echo   '  -u | --usage     show basic usage format'
              echo   '  -v | --verbose   display extra information'
              echo   '       --version   Display version information'
              echo   ''
              echo ' Note: arguments are case sensitive. See documentation for more'
              echo ' detailed information.'
              echo ;;
          *)  echo "Help not avaliable for:$1" ;;
   esac
   exit 0
}

vecho()
{
   if [[ ! $1 = "" ]] ; then
     if $VERBOSE ; then
      echo $1
     fi
   fi
}

have_curl ()
{
  exist=$(which $PROGRAM2)
  if [[ ! $exist = "" ]] ; then
    CHECK_DOCS=true
  else
    echo "Documentation checks requested by could not find [$PROGRAM3]"
    echo "Skipping Checks for documentation references"
  fi
}

PREVIOUS_ARGUMENT=""
do_Assimilate_Arguments(){
   for Arg in $*; do
      case $Arg in
         --help | -h) do_show_help "2"                                ;;
        --usage | -u) do_show_help "1"                                ;;
      --verbose | -v) VERBOSE=true  ; OPT="-vv"                       ;;
           --version) echo `basename $0` "Version $VER" ; exit        ;;
        --quiet | -q) QUIET=true  ; OPT="-q"                          ;;
          --all | -a) FILTER="*"                                      ;;
         --clip | -c) CHECK_CLIB=true                                 ;;
         --docs | -d) have_curl                                       ;;
     --suppress | -s) OPT+=" -u "                                     ;;
    -n | --no_report) REPORT=false                                    ;;
       -r | --report) do_report_only                                  ;;
    	           *) case $PREVIOUS_ARGUMENT in
       --filter | -f) FILTER="$Arg"                                   ;;

                   *)                                                 ;;
                      esac                                            ;;
      esac
      PREVIOUS_ARGUMENT=$Arg
   done
}

do_report_stats() {
  if [ -d $1 ] ; then
    count=(`ls $1/$FILTER -1 2>/dev/null | grep -v ^l | wc -l`)
    echo "Count:$1    Symbols:$count"
  fi
  return $count;
}

do_report_only(){
  local result=0;
  if [ "$FINAL_ARGUMENT" = "" ] ; then
    SRC_DIRS=(`find * -maxdepth 0 -type d -printf "%f "`)
    for subdir in "${SRC_DIRS[@]}" ; do
      do_report_stats $subdir
    done
  else
    if [ -d $FINAL_ARGUMENT ] ; then
      do_report_stats $FINAL_ARGUMENT
      result=$?;
    else
      echo "Error, Could not access [$FINAL_ARGUMENT], recheck"
      result=-1;
    fi
  fi
  exit $result;
}

do_check_docs () {

  local filename;
  local files=$1/${FILTER}
  local result=0;
  local url

  for filename in $files ; do

    fname=`basename $filename`
    vecho "checking documentation for $fname"
    url=$(grep documentation= $filename | cut -f2- -d'=')
    if [ "x$url" != "x" ] ; then
       if ! curl --output /dev/null --location --silent --fail -r 0-0 "$url"; then
         echo "check $url in $fname"
         result=1;
       fi
    else
       echo `basename $filename` "does not appear to have a documentation attribute"
       result=1;
    fi
  done

  return $result;
}

do_check_all_containers () {
  local result=0;
  exist=$(which $PROGRAM1)
  if [[ ! $exist = "" ]] ; then
    SRC_DIRS=(`find * -maxdepth 0 -type d -printf "%f "`)
    for subdir in "${SRC_DIRS[@]}" ; do
      if [ "$(ls -A $subdir/$FILTER 2>/dev/null)" ] ; then
        $PROGRAM1 $OPT $subdir/$FILTER
        if [ $? -ne 0 ] ; then
          result=1;
        fi
      elif [ "$(ls -A $subdir/*/$FILTER 2>/dev/null)" ] ; then
        $PROGRAM1 $OPT $subdir/*/$FILTER
        if [ $? -ne 0 ] ; then
          result=1;
        fi
      else
        echo "$subdir/$FILTER has no symbol files"
      fi
      if $REPORT ; then
        do_report_stats $subdir
      fi
      if $CHECK_DOCS ; then
        do_check_docs $subdir
        if [ $? -ne 0 ] ; then
          result=1;
        fi
      fi
    done
  else
    echo "Error, recheck installation. Could not find [$PROGRAM1]"
  fi
  return $result;
}

do_check_clib_references () {

  local found=0;
  local clibfile=$1;
  local symdir=$2;
  local result=0;

  cd "$symdir"

  SYM_DIRS=(`find * -maxdepth 1 -type d $EXCLUDE -printf "%f "`)

  for subdir in "${SYM_DIRS[@]}" ; do
    vecho "Checking reference to $subdir"
    found=$(cat $clibfile | grep $subdir)
    if [[ -z $found ]] ; then
      missing=$(find * -maxdepth 1 -type d -name "$subdir")
      echo "$missing is not referenced by $filename"
      result=1;
    fi
  done
  return $result;
}

do_check_clib () {

  local result=0;
  local filename;

  exist=$(which $PROGRAM2)
  if [[ ! $exist = "" ]] ; then

    if [ -f "geda-clib.scm" ] ; then
      filename="$PWD/geda-clib.scm"
      result=$($PROGRAM2 "shell" "-e" "(load \"$filename\")" &> /dev/null)
      if [ $? -ne 0 ] ; then
         echo "$filename contains Errors!"
         result=1;
      else
         echo "$filename is Okay!"
         #savedir=$PWD
         cd ..
         #TODO maybe use @symbol_dirs@
         if [ -d "symbols" ] ; then
           do_check_clib_references $filename "symbols"
         elif [ -d "sym" ] ; then
           do_check_clib_references $filename "sym"
         else
           vecho "Not checking for symbol directory references"
         fi
         #cd "$savedir"
      fi
    else
      echo "Error, geda-clib.scm must be in the current directory"
    fi

  else
    echo "Error, recheck installation. Could not find [$PROGRAM2]"
  fi
  return $result;
}

# ----------------------- Parse Command Line ----------------------

do_Assimilate_Arguments $*

if $CHECK_CLIB ; then
  do_check_clib
else
  if [ "x$FINAL_ARGUMENT" != "x" ] && [ -d $FINAL_ARGUMENT ] ; then
    vecho "Checking:$FINAL_ARGUMENT/$FILTER"
    if $QUIET ; then
      $PROGRAM1 $OPT $FINAL_ARGUMENT/${FILTER}
    else
      if $REPORT ; then
        do_report_stats $FINAL_ARGUMENT
      fi
      $PROGRAM1 $OPT $FINAL_ARGUMENT/${FILTER}
      result=$?
    fi
    if $CHECK_DOCS ; then
       do_check_docs $FINAL_ARGUMENT
       if [ $? -ne 0 ] ; then
         result=1;
       fi
    fi
  else
    vecho "Looking in all containers for $FILTER"
    do_check_all_containers
  fi
fi

exit $result
